src/modules/CSADigitizer/CSADigitizerModule.cpp

```diff
@@ -0,0 +1,337 @@
+        int maximum = static_cast<int>(Units::convert(config_.get<int>("output_plots_scale"), "ke"));
+        auto nbins = config_.get<int>("output_plots_bins");
+
+	//asv do these histos make sense? what about one with the noise?
+        // Create histograms if needed
+        h_pxq = new TH1D("pixelcharge", "raw pixel charge;pixel charge [ke];pixels", nbins, 0, maximum);
+        h_tot = new TH1D("tot", "time over threshold;time over threshold [ns];pixels", nbins, 0, maximum);
+        h_toa = new TH1D("toa", "time of arrival;time of arrival [ns];pixels", nbins, 0, maximum);
+	h_pxq_vs_tot = new TH2D("pxqvstot", "ToT vs raw pixel charge;pixel charge [ke];ToT [ns]", nbins, 0, maximum, nbins, 0, tmax_);
+
+    }
+
+    
+}
+
+
+
+void CSADigitizerModule::run(unsigned int event_num) {
+    // Loop through all pixels with charges
+    std::vector<PixelHit> hits;
+    for(auto& pixel_charge : pixel_message_->getData()) {
+        auto pixel = pixel_charge.getPixel();
+        auto pixel_index = pixel.getIndex();
+        auto inputcharge = static_cast<double>(pixel_charge.getCharge());
+
+        LOG(DEBUG) << "Received pixel " << pixel_index << ", charge " << Units::display(inputcharge, "e");
+
+	const auto& pulse = pixel_charge.getPulse(); // the pulse containing charges and times
+	auto pulse_vec = pulse.getPulse();           // the vector of the charges 
+        auto timestep = pulse.getBinning();
+	const int npx = static_cast<int>(ceil(tmax_/timestep));
+
+	if (first_event_){ // initialize impulse response function - assume all time bins are equal
+	  impulseResponse_ = new double[npx];
+	  for (int ipx=0; ipx<npx; ++ipx){
+	    impulseResponse_[ipx]=fImpulseResponse_->Eval(ipx*timestep);
+	  }
+	  first_event_ = false;
+	  LOG(TRACE) << "impulse response initalised. timestep  : " <<  timestep << ", tmax_ : " <<  tmax_ << ", npx " << npx;
+	}
+	
+	Pulse output_pulse(timestep);
+	auto input_length = pulse_vec.size();
+	LOG(TRACE) << "Preparing pulse for pixel " << pixel_index << ", " << pulse_vec.size() << " bins of "
+		   << Units::display(timestep, {"ps", "ns"})
+		   << ", total charge: " << Units::display(pulse.getCharge(), "e");
+	//convolution of the pulse (size input_length) with the impulse response (size npx)
+	for(unsigned long int k=0; k<npx; ++k){
+	  for(unsigned long int i=0; i<=k; ++i){
+	    if( (k-i) < input_length){
+	      //asv to do: not a charge, but voltage pulse... still use this?
+	      output_pulse.addCharge(pulse_vec.at(k-i) * impulseResponse_[i], timestep * static_cast<double>(k));
+	    }
+	  }
+	}
+	auto output_vec = output_pulse.getPulse(); 
+	LOG(TRACE) << "amplified signal" << output_pulse.getCharge()/1e9
+		   << " mV in a pulse with " << output_vec.size() << "bins";
+
+
+	
+	//TOA and TOT logic  
+	auto threshold = config_.get<double>("threshold");
+	bool is_over_threshold=false;
+	double toa, tot;
+	for(long unsigned int i=0; i<output_vec.size(); ++i){
```

Comment:
This here is a bit related to your `first_event_` workaround before, In principle the user should be able to select the clock on which ToA and ToT is counted. The resulting values should be in these clock cycles and not in nanoseconds.

But we might actually do it similar to what you have done now: do the calculation with maximum available precision and convert at the end.

What do you think @annkvth @schuetzepaul

Reply:
I would actually use a range based loop here

```cpp
for (const auto& current : output_vec) {
```

and then use `std::distance` to get your `i`. That's a bit cleaner and easier to read IMHO. It also allows to get rid of your additional check for the end.